File: src/colmap/util/timer_test.cc
// SPDX-License-Identifier: BSD-3-Clause

#include "colmap/util/timer.h"

#include <gtest/gtest.h>

namespace colmap {
namespace {

TEST(Timer, Default) {
  Timer timer;
  EXPECT_EQ(timer.ElapsedMicroSeconds(), 0);
  EXPECT_EQ(timer.ElapsedSeconds(), 0);
  EXPECT_EQ(timer.ElapsedMinutes(), 0);
  EXPECT_EQ(timer.ElapsedHours(), 0);
}

TEST(Timer, Start) {
  Timer timer;
  timer.Start();
  EXPECT_GE(timer.ElapsedMicroSeconds(), 0);
  EXPECT_GE(timer.ElapsedSeconds(), 0);
  EXPECT_GE(timer.ElapsedMinutes(), 0);
  EXPECT_GE(timer.ElapsedHours(), 0);
}

TEST(Timer, Pause) {
  Timer timer;
  timer.Start();
  timer.Pause();
  double prev_time = timer.ElapsedMicroSeconds();
  for (size_t i = 0; i < 1000; ++i) {
    EXPECT_EQ(timer.ElapsedMicroSeconds(), prev_time);
    prev_time = timer.ElapsedMicroSeconds();
  }
  timer.Resume();
  for (size_t i = 0; i < 1000; ++i) {
    EXPECT_GE(timer.ElapsedMicroSeconds(), prev_time);
  }
  timer.Reset();
  EXPECT_EQ(timer.ElapsedMicroSeconds(), 0);
}

}  // namespace
}  // namespace colmap

File: src/pycolmap/util/timer_test.py
import time

import pycolmap


def test_timer_init() -> None:
    timer = pycolmap.Timer()
    assert timer is not None


def test_timer_start() -> None:
    timer = pycolmap.Timer()
    timer.start()


def test_timer_restart() -> None:
    timer = pycolmap.Timer()
    timer.start()
    timer.restart()


def test_timer_pause() -> None:
    timer = pycolmap.Timer()
    timer.start()
    timer.pause()


def test_timer_resume() -> None:
    timer = pycolmap.Timer()
    timer.start()
    timer.pause()
    timer.resume()


def test_timer_reset() -> None:
    timer = pycolmap.Timer()
    timer.start()
    timer.reset()


def test_timer_elapsed_seconds() -> None:
    timer = pycolmap.Timer()
    timer.start()
    time.sleep(0.001)
    elapsed = timer.elapsed_seconds()
    assert elapsed > 0.0


def test_timer_elapsed_micro_seconds() -> None:
    timer = pycolmap.Timer()
    timer.start()
    elapsed = timer.elapsed_micro_seconds()
    assert isinstance(elapsed, float)
    assert elapsed >= 0.0


def test_timer_elapsed_minutes() -> None:
    timer = pycolmap.Timer()
    timer.start()
    elapsed = timer.elapsed_minutes()
    assert isinstance(elapsed, float)
    assert elapsed >= 0.0


def test_timer_elapsed_hours() -> None:
    timer = pycolmap.Timer()
    timer.start()
    elapsed = timer.elapsed_hours()
    assert isinstance(elapsed, float)
    assert elapsed >= 0.0


def test_timer_print_seconds() -> None:
    timer = pycolmap.Timer()
    timer.start()
    timer.print_seconds()


def test_timer_print_minutes() -> None:
    timer = pycolmap.Timer()
    timer.start()
    timer.print_minutes()


def test_timer_print_hours() -> None:
    timer = pycolmap.Timer()
    timer.start()
    timer.print_hours()
